rs/src/parse/plan.rs

```diff
@@ -62,15 +62,109 @@ fn parse_plan_rel(x: &substrait::PlanRel, y: &mut context::Context) -> diagnosti
     Ok(())
 }
 
-/// Toplevel parse function for a plan.
-pub fn parse_plan(x: &substrait::Plan, y: &mut context::Context) -> diagnostic::Result<()> {
+/// Parse a git hash string.
+fn parse_git_hash(x: &String, y: &mut context::Context) -> diagnostic::Result<()> {
+    if !x.is_empty() {
+        static GIT_HASH_RE: once_cell::sync::Lazy<regex::Regex> =
+            once_cell::sync::Lazy::new(|| regex::Regex::new("[0-9a-f]{40}").unwrap());
+        if !GIT_HASH_RE.is_match(x) {
+            diagnostic!(
+                y,
+                Error,
+                IllegalValue,
+                "git hash must be a 40-character lowercase hexadecimal string \
+                if specified."
+            );
+        }
+        diagnostic!(
+            y,
+            Warning,
+            Versioning,
+            "a git hash was specified for the Substrait version, indicating \
+            use of nonstandard features. The validation result may not be \
+            accurate."
+        );
+    }
+    Ok(())
+}
+
+/// Parse a producer identification string.
+fn parse_producer_id(x: &String, y: &mut context::Context) -> diagnostic::Result<()> {
```

Comment:
Same problem as above. I tried briefly to fix the macros at the time to make this work but couldn't immediately figure it out and gave up. I guess I could also write these functions with `T: AsRef<str>` instead of `&String`. That also seems a bit excessive for a private function that is only called with `&String`, though.